### PR TITLE
`xxxForCausalLM` support

### PR DESCRIPTION
**Pull request by younesbelkada**
# What does this PR do?

This PR adds support to all `xxxForCausalLM` models from `transformers`

## TODO

- [x] testing suite
- [x] Push tiny models on https://huggingface.co/trl-internal-testing 

Draft for now 
cc @lvwerra @edbeeching

### CODE REVIEW

File: trl/models/modeling_vhead.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+        if hasattr(config, "summary_activation") and config.summary_activation == "tanh":
+            self.activation = nn.Tanh()
+
+        self.first_dropout = nn.Identity()
+        if hasattr(config, "summary_first_dropout") and config.summary_first_dropout > 0:
+            self.first_dropout = nn.Dropout(config.summary_first_dropout)
+
+        self.last_dropout = nn.Identity()
+        if hasattr(config, "summary_last_dropout") and config.summary_last_dropout > 0:
+            self.last_dropout = nn.Dropout(config.summary_last_dropout)
```
Comment (younesbelkada) [inline trl/models/modeling_vhead.py:48]: This needs to be simplified in the future

Comment (younesbelkada) [reply trl/models/modeling_vhead.py:48]: From what I can see, we'll probably need an extra arg (`v_head_activation`) and simplify this, how does that sound @edbeeching @lvwerra ?

Comment (lvwerra) [reply trl/models/modeling_vhead.py:48]: I guess the `v_head` never has an activation but maybe @edbeeching knows better so activation is actually not need at all. Also not clear if we ever want to apply dropout to the output (`last_dropout`).

Comment (lvwerra) [reply trl/models/modeling_vhead.py:48]: also all the summary attributes are not needed as far as I can tell. maybe all we need is dropout and then a linear layer. can you check what the defaults are?

File: tests/utils/testing_utils_common.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
+import tempfile
+import torch
+
+class BaseModelTester:
+    all_model_names = None
+    trl_model_class = None
+
+    def test_from_save(self):
```
Comment (lvwerra) [inline tests/utils/testing_utils_common.py:8]: another useful test would be if we can load a LM with a v head, save it and then load it as a pure LM. that's what most people will likely use for inference.

File: tests/test_modeling_vhead.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
+import unittest
```
Comment (lvwerra) [inline tests/test_modeling_vhead.py:1]: I think we can just call this the `test_modeling.py` file and add the `BaseModelTester`.

**Review by lvwerra — COMMENTED** [review]
Looks really good @younesbelkada! Just a few minor comments and nits and then this is good to go!

File: trl/models/modeling_vhead.py (hunk 1 above)
Comment (younesbelkada) [reply trl/models/modeling_vhead.py:48]: Here are the defaults: 
```
(v_head): ValueHead(
    (summary): Linear(in_features=768, out_features=2, bias=True)
    (activation): Identity()
    (first_dropout): Dropout(p=0.1, inplace=False)
    (last_dropout): Identity()
    (flatten): Flatten(start_dim=1, end_dim=-1)
  )
```
Let's stick to this architecture now and address a potential adding of activation etc in a follow up PR

Comment (younesbelkada) [reply trl/models/modeling_vhead.py]: also it seems that `flatten` is never used, I removed it

File: tests/utils/testing_utils_common.py (hunk 2 above)
Comment (younesbelkada) [reply tests/utils/testing_utils_common.py]: Great catch! 
Added a test `test_from_save_transformers`, you can now train a model with `trl` save it, and load it back with `transformers`, ins't that cool? :D

File: tests/test_modeling_vhead.py (hunk 3 above)
Comment (younesbelkada) [reply tests/test_modeling_vhead.py]: Should be addressed now ;)

**Comment by younesbelkada** [comment]
Seems to be converging with the latest changes: https://wandb.ai/distill-bloom/gpt2-test/runs/1sxufahx?workspace=user-younesbelkada

File: trl/models/modeling_vhead.py (hunk 1 above)
Comment (edbeeching) [reply trl/models/modeling_vhead.py]: I have never seen a vhead with an activation. You can imagine situations that might use one, but they are so rare that I agree this is not needed.

Comment (younesbelkada) [reply trl/models/modeling_vhead.py]: Perfect thank you for confirming !!

File: trl/ppo.py [hunk 4]
```diff
@@ -260,6 +260,9 @@ def loss(self, old_logprobs, values, rewards, query, response, model_input):
 
         ratio = torch.exp(logprob - old_logprobs)
 
+        if len(ratio.shape) == 2:
```
Comment (lvwerra) [inline trl/ppo.py:263]: Why is that step needed now?

File: examples/scripts/04-ppo-sentiment.py [hunk 5]
```diff
@@ -55,8 +55,8 @@
 
 sentiment_pipe = pipeline("sentiment-analysis","lvwerra/distilbert-imdb", device=pipe_device)
 
-gpt2_model = GPT2HeadWithValueModel.from_pretrained(config['model_name'])
-gpt2_model_ref = GPT2HeadWithValueModel.from_pretrained(config['model_name'])
+gpt2_model = AutoModelForCausalLMWithValueHead.from_pretrained(config['model_name'], summary_dropout_prob=0.1)
```
Comment (lvwerra) [inline examples/scripts/04-ppo-sentiment.py:58]: shouldn't the dropout still be the same?

**Review by lvwerra — APPROVED** [review]
Just two small questions otherwise looks good to me :)

File: examples/scripts/04-ppo-sentiment.py (hunk 5 above)
Comment (younesbelkada) [reply examples/scripts/04-ppo-sentiment.py:58]: By default the dropout is set to `None` (Identity) --> https://github.com/younesbelkada/trl/blob/8390e3bfabdb70da558361483981199304095f05/trl/models/modeling_vhead.py#L26
I think we have to set it to `0.1` since by default it was set to `0.1` on the previous script

File: trl/ppo.py (hunk 4 above)
Comment (younesbelkada) [reply trl/ppo.py:263]: using the latest version of `transformers`, it seems that `ratio` needs a thrid dimension otherwise the broadcasting op wont work when calling `*` in `pg_losses = -advantages * ratio` (for instance using gpt2,  `advantages.shape==(1, 9, 2)` and `ratio.shape == (1, 9)`). 
I got this issue with the latest versions of `transformers`, so I suspect that the way the outputs are dealt changed between the versions but didn't investigated further

File: examples/scripts/04-ppo-sentiment.py (hunk 5 above)
Comment (younesbelkada) [reply examples/scripts/04-ppo-sentiment.py]: As discussed, let's set the dropout to `0.1` by default , [cf77820](https://github.com/lvwerra/trl/pull/53/commits/cf778206b734d9cefb79e318a4fb936a92c053e5)
